tests/test_advanced_streaming.py

```diff
@@ -0,0 +1,142 @@
+"""
+Advanced streaming patterns using bidirectional streams and contexts.
+
+"""
+import itertools
+from typing import Set, Dict, List
+
+import trio
+import tractor
+
+
+_registry: Dict[str, Set[tractor.ReceiveMsgStream]] = {
```

Comment:
Linking comment https://github.com/goodboy/tractor/commit/f9dd2add43a478eeb9a6f8a8d852c23953553e5f#r50700925.

Reply:
@gc-ss mentioned,

> … but if you don't like that, what if we make _registry an argument to publisher?

Yup can do that, I just didn't bc turns out the consumers can update via sending updates on their individual streams.

> It would be perfect if publisher accepted an environment variable (or a Context?)

This can also be done though I'm not sure what you mean by *environment*. A `Context` isn't really required explicitly here but could be used to control cancellation if wanted.

For an alt to `_registry` we could do something like,

```python
class PublisherState:
    subscriptions: dict[str, tractor.MsgStream] = {}
```
kinda thing?

Reply:
> I'm not sure what you mean by environment. A Context isn't really required explicitly here but could be used to control cancellation if wanted

By environment or Context, I meant the context/environment/configuration due to which the function might behave in a non-deterministic manner but the values of which don't necessarily change between function invocations - like a remote host address:port or username/password.

In this case - think about two actors communicating with each other. The exact same actors (types) might behave in very different ways if their context/environment/configuration were different with everything else (eg: arguments) being the same.

Unlike arguments though, their context/environment/configuration dont change between function invocations (like arguments do).

For example, the registry does not need to change between function invocations.

Reply:
I mean, yeah a service actor can easily create wtv objects they need and mutate them as actor-local state.
They can go further and offer a seperate api to mutate that object/state from other actors if needed.

I'm just not sure we need to show that in *this* example since the point is just a super basic dynamic pubsub system.
We do have a [pretty terrible example](https://tractor.readthedocs.io/en/latest/#actor-local-aka-process-global-variables) in the docs that could be improved.

Bonus points for sweet PRs 😉